libflux/src/semantic/walk.rs

```diff
@@ -0,0 +1,1131 @@
+            Node::TextPart(n) => &n.loc,
+            Node::InterpolatedPart(n) => &n.loc,
+            Node::VariableAssgn(n) => &n.loc,
+            Node::MemberAssgn(n) => &n.loc,
+        }
+    }
+}
+
+// Util private functions for node conversion and duplication.
+impl<'a> Node<'a> {
+    fn from_expr(expr: &'a mut Expression) -> Node {
+        match *expr {
+            Expression::Identifier(ref mut e) => Node::IdentifierExpr(e),
+            Expression::Array(ref mut e) => Node::ArrayExpr(e),
+            Expression::Function(ref mut e) => Node::FunctionExpr(e),
+            Expression::Logical(ref mut e) => Node::LogicalExpr(e),
+            Expression::Object(ref mut e) => Node::ObjectExpr(e),
+            Expression::Member(ref mut e) => Node::MemberExpr(e),
+            Expression::Index(ref mut e) => Node::IndexExpr(e),
+            Expression::Binary(ref mut e) => Node::BinaryExpr(e),
+            Expression::Unary(ref mut e) => Node::UnaryExpr(e),
+            Expression::Call(ref mut e) => Node::CallExpr(e),
+            Expression::Conditional(ref mut e) => Node::ConditionalExpr(e),
+            Expression::StringExpr(ref mut e) => Node::StringExpr(e),
+            Expression::Integer(ref mut e) => Node::IntegerLit(e),
+            Expression::Float(ref mut e) => Node::FloatLit(e),
+            Expression::StringLit(ref mut e) => Node::StringLit(e),
+            Expression::Duration(ref mut e) => Node::DurationLit(e),
+            Expression::Uint(ref mut e) => Node::UintLit(e),
+            Expression::Boolean(ref mut e) => Node::BooleanLit(e),
+            Expression::DateTime(ref mut e) => Node::DateTimeLit(e),
+            Expression::Regexp(ref mut e) => Node::RegexpLit(e),
+        }
+    }
+    fn from_stmt(stmt: &'a mut Statement) -> Node {
+        match *stmt {
+            Statement::Expr(ref mut s) => Node::ExprStmt(s),
+            Statement::Variable(ref mut s) => Node::VariableAssgn(s),
+            Statement::Option(ref mut s) => Node::OptionStmt(s),
+            Statement::Return(ref mut s) => Node::ReturnStmt(s),
+            Statement::Test(ref mut s) => Node::TestStmt(s),
+            Statement::Builtin(ref mut s) => Node::BuiltinStmt(s),
+        }
+    }
+    fn from_string_expr_part(sp: &'a mut StringExprPart) -> Node {
+        match *sp {
+            StringExprPart::Text(ref mut t) => Node::TextPart(t),
+            StringExprPart::Interpolated(ref mut e) => Node::InterpolatedPart(e),
+        }
+    }
+    fn from_assignment(a: &'a mut Assignment) -> Node {
+        match *a {
+            Assignment::Variable(ref mut v) => Node::VariableAssgn(v),
+            Assignment::Member(ref mut m) => Node::MemberAssgn(m),
+        }
+    }
+    // Duplicate returns a new Node that points to the same semantic node of `self`.
+    // It is not named "copy" or "clone" to avoid collision with those traits.
+    // This is required to let `visit` own the node.
+    fn duplicate<'b>(&'b mut self) -> Node<'b> {
```

Comment:
I haven't found a better way of doing this. I tried to implement `Clone` but I had problems with lifetimes. This is the best thing I could do. This gives the great advantage that `visit` calls can own nodes.

Reply:
Is this better than using Rc<Node> in your opinion?

Reply:
Yeah, sure. There is and intermediate step, we could use `&mut Node`.

Reply:
We don't actually need multiple ownership, a reference to a mutable variable should be enough.